File: backend/src/models/copper_mine.py
"""
Copper Mine Pydantic Models

Models for copper mine management (LIFF + Dashboard).

符合 CLAUDE.md 🔴:
- Pydantic V2 syntax (ConfigDict, from_attributes)
- Type hints for all fields
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# =============================================================================
# Type Aliases
# =============================================================================

AllowedLevel = Literal["nine", "ten", "both"]

# =============================================================================
# Copper Mine Entity Models
# =============================================================================


class CopperMine(BaseModel):
    """Copper mine entity (unified for LIFF and Dashboard)"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    alliance_id: UUID
    registered_by_line_user_id: str
    game_id: str
    coord_x: int
    coord_y: int
    level: int
    status: str = "active"
    notes: str | None = None
    # New fields for Dashboard integration
    season_id: UUID | None = None
    member_id: UUID | None = None
    registered_at: datetime
    updated_at: datetime


# =============================================================================
# Request Models (from LIFF)
# =============================================================================


class CopperMineCreate(BaseModel):
    """Request to register a copper mine (from LIFF)"""

    line_group_id: str = Field(..., alias="groupId")
    line_user_id: str = Field(..., alias="userId")
    game_id: str = Field(..., alias="gameId", min_length=1, max_length=100)
    coord_x: int = Field(..., alias="coordX", ge=0)
    coord_y: int = Field(..., alias="coordY", ge=0)
    level: int = Field(..., ge=1, le=10)
    notes: str | None = Field(None, max_length=500)

    model_config = ConfigDict(populate_by_name=True)


# =============================================================================
# Response Models
# =============================================================================


class CopperMineResponse(BaseModel):
    """Copper mine for LIFF display"""

    id: str
    game_id: str
    coord_x: int
    coord_y: int
    level: int
    status: str
    notes: str | None = None
    registered_at: datetime


class CopperMineListResponse(BaseModel):
    """Response for copper mine list query"""

    mines: list[CopperMineResponse] = []
    total: int = 0
    my_count: int = 0
    max_allowed: int = 0


class RegisterCopperResponse(BaseModel):
    """Response after registering copper mine"""

    success: bool
    mine: CopperMineResponse | None = None
    message: str | None = None


# =============================================================================
# Dashboard Response Models (with joined member data)
# =============================================================================


class CopperMineOwnershipResponse(BaseModel):
    """Copper mine ownership for Dashboard display (with member info)"""

    id: str
    season_id: str
    member_id: str | None
    coord_x: int
    coord_y: int
    level: int
    applied_at: datetime
    created_at: datetime
    # P1 修復: 添加註冊來源欄位
    registered_via: Literal["liff", "dashboard"] = "dashboard"
    # Joined fields
    member_name: str
    member_group: str | None = None
    line_display_name: str | None = None


class CopperMineOwnershipListResponse(BaseModel):
    """Response for ownership list query"""

    ownerships: list[CopperMineOwnershipResponse] = []
    total: int = 0


# =============================================================================
# Copper Mine Rules Models
# =============================================================================


class CopperMineRule(BaseModel):
    """Copper mine rule entity"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    alliance_id: UUID
    tier: int
    required_merit: int
    allowed_level: AllowedLevel
    created_at: datetime
    updated_at: datetime


class CopperMineRuleCreate(BaseModel):
    """Request to create a copper mine rule"""

    tier: int = Field(..., ge=1, le=10)
    required_merit: int = Field(..., gt=0)
    allowed_level: AllowedLevel = "both"


class CopperMineRuleUpdate(BaseModel):
    """Request to update a copper mine rule"""

    required_merit: int | None = Field(None, gt=0)
    allowed_level: AllowedLevel | None = None


class CopperMineRuleResponse(BaseModel):
    """Copper mine rule response"""

    id: str
    alliance_id: str
    tier: int
    required_merit: int
    allowed_level: AllowedLevel
    created_at: datetime
    updated_at: datetime


# =============================================================================
# Dashboard Request Models
# =============================================================================


class CopperMineOwnershipCreate(BaseModel):
    """Request to create a copper mine ownership (from Dashboard)"""

    member_id: str = Field(..., description="Member UUID")
    coord_x: int = Field(..., ge=0)
    coord_y: int = Field(..., ge=0)
    level: int = Field(..., ge=9, le=10)
    applied_at: datetime | None = None

File: frontend/src/types/copper-mine.ts
/**
 * Copper Mine Management TypeScript Types
 *
 * Type definitions for copper mine rules and ownership records.
 * All field names use snake_case to match API response.
 */

// =============================================================================
// Rule Types (Alliance Level)
// =============================================================================

/**
 * Allowed copper mine levels for a rule tier
 * - nine: Only level 9 allowed
 * - ten: Only level 10 allowed
 * - both: Either level 9 or 10 allowed
 */
export type AllowedLevel = 'nine' | 'ten' | 'both'

/**
 * Copper mine application rule for a specific tier
 */
export interface CopperMineRule {
  readonly id: string
  readonly alliance_id: string
  readonly tier: number
  readonly required_merit: number
  readonly allowed_level: AllowedLevel
  readonly created_at: string
  readonly updated_at: string
}

export interface CreateCopperMineRuleRequest {
  readonly tier: number
  readonly required_merit: number
  readonly allowed_level: AllowedLevel
}

export interface UpdateCopperMineRuleRequest {
  readonly required_merit?: number
  readonly allowed_level?: AllowedLevel
}

// =============================================================================
// Ownership Types (Season Level)
// =============================================================================

/**
 * Copper mine ownership record with joined member data
 */
export interface CopperMineOwnership {
  readonly id: string
  readonly season_id: string
  readonly member_id: string
  readonly coord_x: number
  readonly coord_y: number
  readonly level: 9 | 10
  readonly applied_at: string
  readonly created_at: string
  // Joined fields from members table
  readonly member_name: string
  readonly member_group: string | null
  // Joined field from member_line_bindings (may be null)
  readonly line_display_name: string | null
}

export interface CreateCopperMineOwnershipRequest {
  readonly member_id: string
  readonly coord_x: number
  readonly coord_y: number
  readonly level: 9 | 10
  readonly applied_at?: string
}

// =============================================================================
// Member Status Types (For Validation)
// =============================================================================

/**
 * Member's copper mine application status for validation
 */
export interface MemberCopperMineStatus {
  readonly member_id: string
  readonly member_name: string
  readonly current_count: number
  readonly total_merit: number
  readonly next_tier: number | null
  readonly next_required_merit: number | null
  readonly next_allowed_level: AllowedLevel | null
  readonly can_apply: boolean
}

// =============================================================================
// Display Helpers
// =============================================================================

export const ALLOWED_LEVEL_LABELS: Record<AllowedLevel, string> = {
  nine: '9 級',
  ten: '10 級',
  both: '9 或 10 級',
}

